omniledger/service/service.go

```diff
@@ -80,163 +85,181 @@ func (s *Service) CreateSkipchain(req *CreateSkipchain) (
 		return nil, errors.New("version mismatch")
 	}
 
-	kp := key.NewKeyPair(cothority.Suite)
-
-	tmpColl := collection.New(collection.Data{}, collection.Data{})
-	sigBuf, err := network.Marshal(&req.Transaction.Signature)
-	if err != nil {
-		return nil, errors.New("Couldn't marshal Signature: " + err.Error())
-	}
-	err = tmpColl.Add(req.Transaction.Key, req.Transaction.Value, sigBuf)
-	if err != nil {
-		return nil, errors.New("error while storing in collection: " + err.Error())
-	}
-
-	mr := tmpColl.GetRoot()
-	data := &Data{
-		MerkleRoot:   mr,
-		Transactions: []*Transaction{&req.Transaction},
-		Timestamp:    time.Now().Unix(),
-	}
-
-	buf, err := network.Marshal(data)
-	if err != nil {
-		return nil, errors.New("Couldn't marshal data: " + err.Error())
-	}
-
-	var genesisBlock = skipchain.NewSkipBlock()
-	genesisBlock.Data = buf
-	genesisBlock.Roster = &req.Roster
-	genesisBlock.MaximumHeight = 1
-	genesisBlock.BaseHeight = 1
-
-	// TODO: Signature?
-	var ssb = skipchain.StoreSkipBlock{NewBlock: genesisBlock}
-	ssbReply, err := s.skService().StoreSkipBlock(&ssb)
+	sb, err := s.createNewBlock(nil, &req.Roster, []Transaction{req.Transaction})
 	if err != nil {
 		return nil, err
 	}
-	skID := ssbReply.Latest.SkipChainID()
-	gid := string(skID)
-
-	err = s.getCollection(skID).Store(req.Transaction.Key, req.Transaction.Value, sigBuf)
-	if err != nil {
-		return nil, errors.New(
-			"error while storing in collection: " + err.Error())
-	}
-	s.storage.Private[gid] = kp.Private
 	s.save()
 	return &CreateSkipchainResponse{
 		Version:   CurrentVersion,
-		Skipblock: ssbReply.Latest,
+		Skipblock: sb,
 	}, nil
 }
 
 // SetKeyValue asks cisc to add a new key/value pair.
 func (s *Service) SetKeyValue(req *SetKeyValue) (*SetKeyValueResponse, error) {
-	// Check the input arguments
-	// TODO: verify the signature on the key/value pair
 	if req.Version != CurrentVersion {
 		return nil, errors.New("version mismatch")
 	}
-	gid := string(req.SkipchainID)
-	latest, err := s.db().GetLatest(s.db().GetByID(req.SkipchainID))
+
+	s.queuesMutex.Lock()
+	defer s.queuesMutex.Unlock()
+	ssid := string(req.SkipchainID)
+	if s.queues[ssid] == nil {
+		s.queues[ssid] = &queue{transactions: []Transaction{req.Transaction}}
+		go func() {
+			<-time.After(waitQueueing)
+			s.queuesMutex.Lock()
+			q := s.queues[ssid]
+			s.queues[ssid] = nil
+			s.queuesMutex.Unlock()
+			sb, err := s.db().GetLatest(s.db().GetByID(req.SkipchainID))
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			log.Lvlf2("Creating block with transactions %+v", q.transactions)
+			_, err = s.createNewBlock(req.SkipchainID, sb.Roster, q.transactions)
+			if err != nil {
+				log.Error("couldn't create new block: " + err.Error())
+				return
+			}
+		}()
+	} else {
+		s.queues[ssid].transactions = append(s.queues[ssid].transactions, req.Transaction)
+	}
```

Comment:
the above logic works, but doesn't feel very idiomatic. I'd use a long-running go-routine or even one go-routine for every `ssid`, and then send transactions to the appropriate channel. Each go-routine would have like a `select <-time.After(t)` and creates a new block. _"Don't communicate by sharing memory, share memory by communicating."_

Reply:
Yes, I thought about that when writing it - but then we would need to have one worker for every skipchain that is initialised. Is that OK?

Reply:
And we'll have dangling go-routines in the tests :(

Reply:
If you can have a look and tell me if that makes sense to you - specifically the need to close the go-routines in the tests...